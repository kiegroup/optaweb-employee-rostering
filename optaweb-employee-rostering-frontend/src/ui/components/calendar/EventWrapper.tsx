/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';

import 'react-big-calendar/lib/css/react-big-calendar.css';
import './ReactBigCalendarOverrides.css';
import { Popover } from '@patternfly/react-core';

// Workaround for https://github.com/intljusticemission/react-big-calendar/issues/1397,
// and helps with styling the event component and allows the popup to show even if you click
// the time (necessary for small events)
export default function EventWrapper(props: React.PropsWithChildren<{
  continuesEarlier: boolean;
  continuesLater: boolean;
  className: string;
  boundary: React.MutableRefObject<HTMLElement | undefined>;
  popoverHeader: React.ReactNode;
  popoverBody: React.ReactNode;
  style?: React.CSSProperties;
}>): JSX.Element {
  let { className } = props;
  const style: React.CSSProperties = {
    backgroundColor: 'transparent',
    border: 'none',
    ...props.style,
  };

  if (props.style !== undefined && props.style.top !== undefined && props.style.height !== undefined) {
    if (props.continuesEarlier) {
      className += ' continues-from-previous-day';
    }
    if (props.continuesLater) {
      className += ' continues-next-day';
    }
  }

  // zIndex need to be higher than .rbc-event:hover:not(.rbc-addons-dnd-drag-preview)
  // in ReactBigCalendarOverrides.css
  return (
    <div
      className={className}
      style={style}
    >
      <Popover
        headerContent={props.popoverHeader}
        bodyContent={props.popoverBody}
        appendTo={props.boundary.current}
        zIndex={1000001}
      >
        {props.children as React.ReactElement}
      </Popover>
    </div>
  );
}
