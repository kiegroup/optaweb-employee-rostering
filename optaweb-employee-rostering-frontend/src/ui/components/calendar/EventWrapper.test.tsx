import * as React from 'react';
import { shallow } from 'enzyme';
import EventWrapper from './EventWrapper';

describe('EventWrapper', () => {
  it('should render correctly when the event does not span multiple days', () => {
    const eventWrapper = shallow(
      <EventWrapper
        continuesEarlier={false}
        continuesLater={false}
        className="class"
        popoverHeader="Title"
        popoverBody="Body"
        boundary={{ current: undefined }}
        style={{ top: '50%', height: '25%', color: 'white' }}
      />,
    );
    expect(eventWrapper).toMatchSnapshot();
  });

  it('should render correctly when the event continues from earlier', () => {
    const eventWrapper = shallow(
      <EventWrapper
        continuesEarlier
        continuesLater={false}
        className="class"
        popoverHeader="Title"
        popoverBody="Body"
        boundary={{ current: undefined }}
        style={{ top: '0%', height: '25%', color: 'white' }}
      />,
    );
    expect(eventWrapper).toMatchSnapshot();
  });

  it('should render correctly when the event continues later', () => {
    const eventWrapper = shallow(
      <EventWrapper
        continuesEarlier={false}
        continuesLater
        className="class"
        popoverHeader="Title"
        popoverBody="Body"
        boundary={{ current: undefined }}
        style={{ top: '75%', height: '25%', color: 'white' }}
      />,
    );
    expect(eventWrapper).toMatchSnapshot();
  });

  it('should render correctly when the event continues both before and after', () => {
    const eventWrapper = shallow(
      <EventWrapper
        continuesEarlier
        continuesLater
        className="class"
        popoverHeader="Title"
        popoverBody="Body"
        boundary={{ current: undefined }}
        style={{ top: '0%', height: '100%', color: 'white' }}
      />,
    );
    expect(eventWrapper).toMatchSnapshot();
  });

  it('should render correctly when the event has no height or top', () => {
    const eventWrapper = shallow(
      <EventWrapper
        continuesEarlier={false}
        continuesLater={false}
        className="class"
        popoverHeader="Title"
        popoverBody="Body"
        boundary={{ current: undefined }}
        style={{ color: 'white' }}
      />,
    );
    expect(eventWrapper).toMatchSnapshot();
  });

  it('should render correctly without a style', () => {
    const eventWrapper = shallow(
      <EventWrapper
        continuesEarlier={false}
        continuesLater={false}
        className="class"
        popoverHeader="Title"
        popoverBody="Body"
        boundary={{ current: undefined }}
      />,
    );
    expect(eventWrapper).toMatchSnapshot();
  });

  it('should render correctly when the event does not have a class name', () => {
    const eventWrapper = shallow(
      <EventWrapper
        continuesEarlier={false}
        continuesLater={false}
        className=""
        popoverHeader="Title"
        popoverBody="Body"
        boundary={{ current: undefined }}
        style={{ top: '50%', height: '25%', color: 'white' }}
      />,
    );
    expect(eventWrapper).toMatchSnapshot();
  });
});
