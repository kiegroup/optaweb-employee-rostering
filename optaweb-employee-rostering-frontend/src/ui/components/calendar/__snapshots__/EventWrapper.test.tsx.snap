// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventWrapper should render correctly when the event continues both before and after 1`] = `
<div
  className="class continues-from-previous-day continues-next-day"
  style={
    Object {
      "backgroundColor": "transparent",
      "border": "none",
      "color": "white",
      "height": "100%",
      "top": "0%",
    }
  }
>
  <Popover
    bodyContent="Body"
    headerContent="Title"
    zIndex={1000001}
  />
</div>
`;

exports[`EventWrapper should render correctly when the event continues from earlier 1`] = `
<div
  className="class continues-from-previous-day"
  style={
    Object {
      "backgroundColor": "transparent",
      "border": "none",
      "color": "white",
      "height": "25%",
      "top": "0%",
    }
  }
>
  <Popover
    bodyContent="Body"
    headerContent="Title"
    zIndex={1000001}
  />
</div>
`;

exports[`EventWrapper should render correctly when the event continues later 1`] = `
<div
  className="class continues-next-day"
  style={
    Object {
      "backgroundColor": "transparent",
      "border": "none",
      "color": "white",
      "height": "25%",
      "top": "75%",
    }
  }
>
  <Popover
    bodyContent="Body"
    headerContent="Title"
    zIndex={1000001}
  />
</div>
`;

exports[`EventWrapper should render correctly when the event does not have a class name 1`] = `
<div
  className=""
  style={
    Object {
      "backgroundColor": "transparent",
      "border": "none",
      "color": "white",
      "height": "25%",
      "top": "50%",
    }
  }
>
  <Popover
    bodyContent="Body"
    headerContent="Title"
    zIndex={1000001}
  />
</div>
`;

exports[`EventWrapper should render correctly when the event does not span multiple days 1`] = `
<div
  className="class"
  style={
    Object {
      "backgroundColor": "transparent",
      "border": "none",
      "color": "white",
      "height": "25%",
      "top": "50%",
    }
  }
>
  <Popover
    bodyContent="Body"
    headerContent="Title"
    zIndex={1000001}
  />
</div>
`;

exports[`EventWrapper should render correctly when the event has no height or top 1`] = `
<div
  className="class"
  style={
    Object {
      "backgroundColor": "transparent",
      "border": "none",
      "color": "white",
    }
  }
>
  <Popover
    bodyContent="Body"
    headerContent="Title"
    zIndex={1000001}
  />
</div>
`;

exports[`EventWrapper should render correctly without a style 1`] = `
<div
  className="class"
  style={
    Object {
      "backgroundColor": "transparent",
      "border": "none",
    }
  }
>
  <Popover
    bodyContent="Body"
    headerContent="Title"
    zIndex={1000001}
  />
</div>
`;
